### PR TITLE
Upgrade: Espree 2.0.0 and escope 3.0.0 (fixes #2234, fixes #2201, fixes

### LINKED ISSUE #2201: no-unused-vars on default parameters
This may be related to the eslint-babel parser. Given [this code](http://babeljs.io/repl/#?experimental=true&playground=true&evaluate=true&loose=true&spec=false&code=const%20foo%20%3D%20'hello!'%3B%0A%0Afunction%20bar(foobar%20%3D%20foo)%20%7B%0A%20%20console.log(foobar)%3B%0A%7D)

```javascript
// example.js
const foo = "hello!";

function bar(foobar = foo) {
  foobar.replace(/!$/, " world!");
}

bar();
```

and this .eslintrc

```json
{
  "parser": "babel-eslint",
  "rules": {
    "strict": 0
  }
}
```

I get this error:

```
~/Desktop
⛄  $ eslint example.js 

example.js
  1:6  error  foo is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

However, `foo` is being used as the default parameter.

### LINKED ISSUE #2234: Upgrade Espree and escope
I've just finished ESTree changes for Espree, which means I can also attempt to upgrade escope. These changes should solve a lot of our outstanding problems related to scoping, but will likely require a lot of changes in a number of places to fully work.

The upgrades will be to Espree 2.0.0 and escope 3.0.0, and these need to be done together due to the ESTree changes.

I'm going to start work on this today, though it may take several days to finish.

### PR DESCRIPTION
**Pull request by nzakas**
This fixes a bunch of outstanding ES6 scoping issues!

### CODE REVIEW

File: lib/eslint.js [hunk 1]
```diff
@@ -24,19 +24,10 @@ var estraverse = require("estraverse-fb"),
 // Helpers
 //------------------------------------------------------------------------------
 
-// Have to modify estraverse until TryStatement.handlers is removed
-estraverse.VisitorKeys.TryStatement = ["block", "handlers", "finalizer"];
-
 // TODO: Remove when estraverse is updated
-estraverse.Syntax.ExportNamedDeclaration = "ExportNamedDeclaration";
```
Comment (ilyavolodin) [inline lib/eslint.js]: Should those be removed? estraverse hasn't been updated yet, right?

Comment (nzakas) [inline lib/eslint.js]: I forgot to take this out last time I updated estraverse. Just a little housecleaning.

**Comment by nzakas** [comment]
Holding this out until we are ready for a 1.0.0 RC.

**Comment by nzakas** [comment]
Actually, changed my mind. It would be good to get this in sooner. :)

**Pull request merged**